File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import { ValueError } from "../utils/errors/python-errors.js";
import type { Range, UnresolvedRule } from "./grammar-graph-types.js";
import {
  isRuleChar,
  isRuleCharExclude,
  isRuleEnd,
  isRuleRef,
} from "./type-guards.js";

const KEY_TRANSLATION = {
  RuleEnd: 0,
  RuleChar: 1,
  RuleCharExclude: 2,
} as const;

// Matches the spacing of Python's `json.dumps`, which the reference uses here.
const dumps = (value: (number | Range)[]): string =>
  `[${value
    .map((v) => (Array.isArray(v) ? `[${v.join(", ")}]` : `${v}`))
    .join(", ")}]`;

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION.RuleEnd}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION.RuleChar}-${dumps(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION.RuleCharExclude}-${dumps(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `3-${rule.value}`;
  }

  throw new ValueError(`Unknown rule type: ${String(rule)}`);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T005215Z_9a3e20fe/20260910T045026Z_2a40c4a6/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
import json

from .grammar_graph_types import RuleChar, RuleCharExclude, RuleEnd, UnresolvedRule
from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref

KEY_TRANSLATION = {
    RuleEnd: 0,
    RuleChar: 1,
    RuleCharExclude: 2,
}


def get_serialized_rule_key(rule: UnresolvedRule) -> str:
    if is_rule_end(rule):
        return f"{KEY_TRANSLATION[RuleEnd]}"

    if is_rule_char(rule):
        return f"{KEY_TRANSLATION[RuleChar]}-{json.dumps(rule.value)}"

    if is_rule_char_exclude(rule):
        return f"{KEY_TRANSLATION[RuleCharExclude]}-{json.dumps(rule.value)}"

    if is_rule_ref(rule):
        return f"3-{rule.value}"

    raise ValueError(f"Unknown rule type: {rule!r}")
